common/finalizer/finalizer.go

```diff
@@ -0,0 +1,170 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package finalizer
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/internal/goro"
+)
+
+var (
+	FinalizerAlreadyDoneErr = errors.New("finalizer already finalized")
+	FinalizerUnknownIdErr   = errors.New("finalizer callback not found")
+	FinalizerDuplicateIdErr = errors.New("finalizer callback already registered")
+)
+
+type Finalizer struct {
+	logger    log.Logger
+	mu        sync.Mutex  // lock protects access to callbacks map
+	finalized atomic.Bool // indicating if finalizer has been executed
+	callbacks map[string]func(context.Context) error
+}
+
+func NewFinalizer(
+	logger log.Logger,
+) *Finalizer {
+	return &Finalizer{
+		logger:    logger,
+		callbacks: make(map[string]func(context.Context) error),
+	}
+}
+
+// Register adds a callback to the finalizer.
+// Returns an error if the ID is already registered, or when the finalizer already ran.
+func (f *Finalizer) Register(
+	id string,
+	callback func(context.Context) error,
+) error {
+	if f.finalized.Load() {
+		return FinalizerAlreadyDoneErr
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
```

Comment:
there could be a race condition here?

Reply:
Oh, I see it now! The for loop can still iterate over the callbacks after the mutex was unlocked. I'm going to add a step to ensure that the goroutine completes before returning and releasing the lock. Since at that point, the pool has stopped, the calls to `Do` won't take a lot of time anymore. So I think it's safe to wait for that.

Reply:
oh sorry, I was actually referring to something else. I am thinking there's a race condition between Register() and Run() methods. Register() can first pass the finalized check, Run() method then change the flag, complete all the callbacks, then Register adds the new callback into the callbacks list but it will never be invoked.


But thanks for catching the other issue!

Reply:
That's true. And in terms of logic that's fine, once the Finalizer is started, any callbacks that are added after it completes will be ignored. But this prompted me to re-visit this since the call to Register would have to block until the Run completes. Which is not ideal. So I now re-structured it a bit to ensure that Register/Deregister unblock immediately when Run is/was running.

Reply:
👍 yeah I mainly want the finalizer to have a clear contract with the caller and caller can know for sure if the callback registered will be eventually invoked or not.